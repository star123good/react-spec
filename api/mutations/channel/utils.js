// @flow
import createLoaders from '../../loaders/';
const loaders = createLoaders();

export const userCanManageChannel = async (
  userId: string,
  channelId: string
): Promise<boolean> => {
  if (!userId || !channelId) return false;

  const channel = await loaders.channel.load(channelId);

  if (!channel || channel.deletedAt) return false;
  const [communityPermissions, channelPermissions] = await Promise.all([
    loaders.userPermissionsInCommunity.load([userId, channel.communityId]),
    loaders.userPermissionsInChannel.load([userId, channelId]),
  ]);

  if (!communityPermissions) return false;
  if (communityPermissions.isOwner || communityPermissions.isModerator)
    return true;
  if (!channelPermissions) return false;
  if (channelPermissions.isOwner || channelPermissions.isModerator) return true;
  return false;
};

export const userCanCreateChannel = async (
  userId: string,
  communityId: string
): Promise<boolean> => {
  if (!userId || !communityId) return false;
  const permissions = loaders.userPermissionsInCommunity.load([
    userId,
    communityId,
  ]);
  if (!permissions || !permissions.isOwner) return false;
  return true;
};
