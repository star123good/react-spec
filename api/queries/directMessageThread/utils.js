// @flow

export const canViewDMThread = async (
  threadId: string,
  userId: string,
  { loaders }: { loaders: Object }
) => {
  if (!userId) return false;

  const result = await loaders.directMessageParticipants.load(threadId);
  if (!result || !result.reduction || result.reduction.length === 0)
    return false;

  const members = result.reduction;
  const ids = members.map(({ userId }) => userId);
  if (ids.indexOf(userId) === -1) return false;

  return true;
};
