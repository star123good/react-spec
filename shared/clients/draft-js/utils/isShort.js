// @flow
import type { MessageInfoType } from '../../../graphql/fragments/message/messageInfo';

export const isShort = (message: MessageInfoType): boolean => {
  if (message.messageType === 'media') return false;
  const jsonBody = JSON.parse(message.content.body);
  return (
    jsonBody.blocks.length <= 1 &&
    jsonBody.blocks
      .filter(block => block.type === 'unstyled')
      .map(block => block.text).length <= 170
  );
};
