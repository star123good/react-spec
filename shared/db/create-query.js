/* @flow
 *
 * Create a RethinkDB query and cache it's result in a TagCache
 *
 * Usage:
 *
 * const getThreadById = createReadQuery({
 *   // NOTE: No .run() at the end of the query!!!
 *   query: (threadId: string) => db.table('threads').get(threadId),
 *   tags: (threadId: string) => (thread: ?DBThread) => thread ? [thread.id, thread.communityId, thread.channelId] : [],
 * });
 *
 * const updateUser = createWriteQuery({
 *  // NOTE: .run() at the end of the query
 *  query: (userId: string, data: Object) => db.table('users').get(userId).update(data).run(),
 *  invalidateTags: (userId: string) => () => [userId]
 * });
 */

const debug = require('debug')('shared:rethinkdb:db-query-cache');
import queryCache from './query-cache';
import { READ_RUN_ERROR, WRITE_RUN_ERROR } from './constants';

let TOTAL_QUERIES = 0;
let CACHED_RESULTS = 0;

if (debug.enabled) {
  debug('Logging of cache hit rate enabled! (every 30s)');
  setInterval(() => {
    debug(
      `Cache hit rate: ${(CACHED_RESULTS / TOTAL_QUERIES) *
        100}% (${CACHED_RESULTS} cached of ${TOTAL_QUERIES} total)`
    );
  }, 30000);
}

type RethinkDBQuery<O> = {
  toString: Function,
  run: () => Promise<O>,
};

type ProcessFn<I, O> = (data: O) => Promise<*> | *;
type TagsFn<I, O> = (data: O) => Array<?string>;

type CreateReadQueryInput<I, O> = $Exact<{
  query: Promise<RethinkDBQuery<O>> | RethinkDBQuery<O>,
  process?: ProcessFn<I, O>,
  tags: TagsFn<I, O>,
}>;

type CreateQueryCallback<I, O> = (...args: I) => O;

export const createReadQuery = <I: Array<*>, O: *>(
  callback: (...args: I) => CreateReadQueryInput<I, O>
) => {
  return async (...args: I) => {
    const input = callback(...args);
    TOTAL_QUERIES++;
    const query = await input.query;
    if (typeof query.run !== 'function') throw new Error(READ_RUN_ERROR);

    const queryString = query.toString();
    const cached = await queryCache.get(queryString);
    if (cached) {
      CACHED_RESULTS++;
      return cached;
    }

    const result = await query
      .run()
      .then(input.process ? input.process : res => res);

    const tags = input.tags(result).filter(Boolean);
    await queryCache.set(queryString, result, tags);
    return result;
  };
};

type CreateWriteQueryInput<I, O> = $Exact<{
  query: Promise<O> | O,
  invalidateTags: TagsFn<I, O>,
}>;

export const createWriteQuery = <I: Array<*>, O: *>(
  callback: (...args: I) => CreateWriteQueryInput<I, O>
) => {
  return async (...args: I) => {
    const input = callback(...args);
    const result = await input.query;
    if (typeof result.run === 'function') throw new Error(WRITE_RUN_ERROR);

    const tags = input.invalidateTags(result).filter(Boolean);
    await queryCache.invalidate(...tags);
    return result;
  };
};
