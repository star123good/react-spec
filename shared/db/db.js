// @flow
/**
 * Database setup is done here
 */
import fs from 'fs';
import path from 'path';

const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

const DEFAULT_CONFIG = {
  // Connect to the test database when, well, testing
  db: !process.env.TEST_DB ? 'spectrum' : 'testing',
  max:
    process.env.SENTRY_NAME === 'api' || process.env.SENTRY_NAME === 'hyperion'
      ? 60
      : 1, // Maximum number of connections, default is 1000
  buffer: 1, // Minimum number of connections open at any given moment, default is 50
  timeoutGb: 60 * 1000, // How long should an unused connection stick around, default is an hour, this is a minute
};

let ca;

try {
  ca = fs.readFileSync(path.join(process.cwd(), 'cacert'));
} catch (err) {
  if (IS_PROD) throw err;
}

const PRODUCTION_CONFIG = {
  password: process.env.COMPOSE_RETHINKDB_PASSWORD,
  host: process.env.COMPOSE_RETHINKDB_URL,
  port: process.env.COMPOSE_RETHINKDB_PORT,
  ...(ca
    ? {
        ssl: {
          ca,
        },
      }
    : {}),
};

const config = IS_PROD
  ? {
      ...DEFAULT_CONFIG,
      ...PRODUCTION_CONFIG,
    }
  : {
      ...DEFAULT_CONFIG,
    };

var r = require('rethinkdbdash')(config);

// Exit the process on unhealthy db in test env
if (process.env.TEST_DB) {
  r.getPoolMaster().on('healthy', healthy => {
    if (!healthy) {
      process.exit(1);
    }
  });
}

if (process.env.NODE_ENV === 'development' && process.env.TRACK_DB_PERF) {
  const fs = require('fs');
  const inspect = require('rethinkdb-inspector');
  const queries = [];
  inspect(r, {
    onQueryComplete: (query, { size, time }) => {
      if (query.indexOf('.changes') > -1) return;
      queries.push({ query, time, size });
      fs.writeFileSync(
        'queries-by-time.js',
        JSON.stringify(queries.sort((a, b) => b.time - a.time), null, 2)
      );
      fs.writeFileSync(
        'queries-by-response-size.js',
        JSON.stringify(queries.sort((a, b) => b.size - a.size), null, 2)
      );
    },
  });
}

module.exports = { db: r };
