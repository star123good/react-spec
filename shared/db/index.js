// @flow
import { db } from './db';
import { createReadQuery, createWriteQuery } from './create-query';

export { db, createReadQuery, createWriteQuery };
