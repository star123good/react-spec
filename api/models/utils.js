// @flow
import { db } from 'shared/db';

export type Timeframe = 'daily' | 'weekly' | 'monthly' | 'quarterly';

export const parseRange = (timeframe?: Timeframe) => {
  switch (timeframe) {
    case 'daily': {
      return { current: 60 * 60 * 24, previous: 60 * 60 * 24 * 2 };
    }
    case 'weekly': {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
    case 'monthly': {
      return { current: 60 * 60 * 24 * 30, previous: 60 * 60 * 24 * 60 };
    }
    case 'quarterly': {
      return { current: 60 * 60 * 24 * 90, previous: 60 * 60 * 24 * 180 };
    }
    default: {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
  }
};

export const getAu = (range: Timeframe) => {
  const { current } = parseRange(range);
  return db
    .table('users')
    .filter(row =>
      row
        .hasFields('lastSeen')
        .and(row('lastSeen').during(db.now().sub(current), db.now()))
    )
    .count()
    .default(0)
    .run();
};

// prettier-ignore
export const getGrowth = async (table: string, range: Timeframe, field: string, filter: ?mixed) => {
  const { current, previous } = parseRange(range);
  const currentPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(current), db.now()))
    .filter(filter ? filter : '')
    .count()
    .run();

  const prevPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(previous), db.now().sub(current)))
    .filter(filter ? filter : '')
    .count()
    .run();

  const rate = (await (currentPeriodCount - prevPeriodCount)) / prevPeriodCount;
  return {
    currentPeriodCount,
    prevPeriodCount,
    growth: Math.round(rate * 100),
  };
};

export const getCount = (table: string, filter: mixed) => {
  if (filter) {
    return db
      .table(table)
      .filter(filter)
      .count()
      .run();
  }

  return db
    .table(table)
    .count()
    .run();
};

export const getCoreMetrics = () => {
  return db
    .table('coreMetrics')
    .orderBy(db.desc('date'))
    .limit(90)
    .orderBy('date')
    .run();
};
