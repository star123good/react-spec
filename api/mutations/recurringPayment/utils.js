// @flow
require('now-env');
import UserError from '../../utils/UserError';
import { stripe } from 'shared/stripe';

type Err = {
  type: string,
};

export const parseStripeErrors = (err: Err) => {
  switch (err.type) {
    case 'StripeCardError':
      // A declined card error
      return new UserError(err); // => e.g. "Your card's expiration year is invalid."
    case 'RateLimitError':
      // Too many requests made to the API too quickly
      return new UserError(
        'Could not upgrade to Pro at this time, try again later: 1'
      );
    case 'StripeInvalidRequestError':
      // Invalid parameters were supplied to Stripe's API
      return new UserError(
        'Could not upgrade to Pro at this time, try again later: 2'
      );
    case 'StripeAPIError':
      // An error occurred internally with Stripe's API
      return new UserError(
        'Something went wrong at Stripe, try again later: 3'
      );
    case 'StripeConnectionError':
      // Some kind of error occurred during the HTTPS communication
      return new UserError(
        'Something went wrong at Stripe, try again later: 4'
      );
    case 'StripeAuthenticationError':
      // You probably used an incorrect API key
      return new UserError(
        'Something went wrong at Stripe, try again later: 5'
      );
    default:
      // Handle any other types of unexpected errors
      return new UserError('Something went wrong, try again later: 6');
  }
};

export const getStripeCustomer = (customerId: string) => {
  try {
    return stripe.customers.retrieve(customerId);
  } catch (err) {
    return console.error(err) || err;
  }
};

export const createStripeCustomer = (email: string, source: string) => {
  try {
    return stripe.customers.create({
      email,
      source,
    });
  } catch (err) {
    return console.error(err) || err;
  }
};

export const updateStripeCustomer = (
  customer: string,
  email: string,
  source: string
) => {
  try {
    return stripe.customers.update(customer, {
      email,
      source,
    });
  } catch (err) {
    return console.error(err) || err;
  }
};

export const createStripeSubscription = (
  customer: string,
  plan: string,
  quantity: number
) => {
  try {
    return stripe.subscriptions.create({
      customer,
      items: [
        {
          plan,
          quantity: quantity ? quantity : 1, // if no quantity is provided, default to one
        },
      ],
    });
  } catch (err) {
    return console.error(err) || err;
  }
};

export const deleteStripeSubscription = (subscription: string) => {
  try {
    return stripe.subscriptions.del(subscription);
  } catch (err) {
    return console.error(err) || err;
  }
};
