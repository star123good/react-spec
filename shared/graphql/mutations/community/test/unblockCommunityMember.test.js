//@flow
import { request } from '../../../../../iris/test/utils';
import db from 'shared/testing/db';
import data from 'shared/testing/data';

const owner = data.users.find(({ username }) => username === 'mxstbr');
const member = data.users.find(({ username }) => username === 'bryn');
const nonMember = data.users.find(({ username }) => username === 'bad-boy');
const previousMember = data.users.find(
  ({ username }) => username === 'previous-boy'
);
const blockedMember = data.users.find(
  ({ username }) => username === 'blocked-boy'
);

const input = { communityId: data.communities[0].id, userId: blockedMember.id };
const variables = { input };

const dropUsersCommunities = () =>
  db
    .table('usersCommunities')
    .delete()
    .run();

const populateUsersCommunities = () =>
  db
    .table('usersCommunities')
    .insert(...data.usersCommunities)
    .run();

afterAll(
  async () =>
    await dropUsersCommunities().then(
      async () => await populateUsersCommunities()
    )
);

it('should fail if current user is not authenticated', async () => {
  const query = /* GraphQL */ `
    mutation unblockCommunityMember($input: UnblockCommunityMemberInput!) {
      unblockCommunityMember (input: $input)
    }
  `;

  expect.assertions(2);
  const result = await request(query, { variables });
  expect(result.errors).toHaveLength(1);
  expect(result.data.unblockCommunityMember).toEqual(null);
});

it('should fail if current user is not a member of the community', async () => {
  const query = /* GraphQL */ `
    mutation unblockCommunityMember($input: UnblockCommunityMemberInput!) {
      unblockCommunityMember (input: $input)
    }
  `;

  const context = {
    user: nonMember,
  };

  expect.assertions(2);
  const result = await request(query, { context, variables });
  expect(result.errors).toHaveLength(1);
  expect(result.data.unblockCommunityMember).toEqual(null);
});

it('should fail if evaluated user is not a member of the community', async () => {
  const query = /* GraphQL */ `
    mutation unblockCommunityMember($input: UnblockCommunityMemberInput!) {
      unblockCommunityMember (input: $input)
    }
  `;

  const context = {
    user: owner,
  };

  const thisInput = {
    communityId: data.communities[0].id,
    userId: nonMember.id,
  };
  const thisVariables = { input: thisInput };

  expect.assertions(2);
  const result = await request(query, { context, variables: thisVariables });
  expect(result.errors).toHaveLength(1);
  expect(result.data.unblockCommunityMember).toEqual(null);
});

it('should fail if evaluated user used to be a member but is not any more', async () => {
  const query = /* GraphQL */ `
    mutation unblockCommunityMember($input: UnblockCommunityMemberInput!) {
      unblockCommunityMember (input: $input)
    }
  `;

  const context = {
    user: owner,
  };

  const thisInput = {
    communityId: data.communities[0].id,
    userId: previousMember.id,
  };
  const thisVariables = { input: thisInput };

  expect.assertions(2);
  const result = await request(query, { context, variables: thisVariables });
  expect(result.errors).toHaveLength(1);
  expect(result.data.unblockCommunityMember).toEqual(null);
});

it("should fail if the community doesn't exist", async () => {
  const query = /* GraphQL */ `
    mutation unblockCommunityMember($input: UnblockCommunityMemberInput!) {
      unblockCommunityMember (input: $input)
    }
  `;

  const thisInput = { communityId: 'blah', userId: member.id };
  const thisVariables = { input: thisInput };

  const context = {
    user: owner,
  };

  expect.assertions(2);
  const result = await request(query, { context, variables: thisVariables });
  expect(result.errors).toHaveLength(1);
  expect(result.data.unblockCommunityMember).toEqual(null);
});

it('should fail if evaluated user is not blocked in the community', async () => {
  const query = /* GraphQL */ `
    mutation unblockCommunityMember($input: UnblockCommunityMemberInput!) {
      unblockCommunityMember (input: $input)
    }
  `;

  const thisInput = {
    communityId: data.communities[0].id,
    userId: member.id,
  };
  const thisVariables = { input: thisInput };

  const context = {
    user: owner,
  };

  expect.assertions(2);
  const result = await request(query, { context, variables: thisVariables });
  expect(result.errors).toHaveLength(1);
  expect(result.data.unblockCommunityMember).toEqual(null);
});

it('should fail if current user is not an owner of the community', async () => {
  const query = /* GraphQL */ `
    mutation unblockCommunityMember($input: UnblockCommunityMemberInput!) {
      unblockCommunityMember (input: $input)
    }
  `;

  const context = {
    user: member,
  };

  expect.assertions(2);
  const result = await request(query, { context, variables });
  expect(result.errors).toHaveLength(1);
  expect(result.data.unblockCommunityMember).toEqual(null);
});

it('should block a member in the community', async () => {
  const query = /* GraphQL */ `
    mutation unblockCommunityMember($input: UnblockCommunityMemberInput!) {
      unblockCommunityMember (input: $input)
    }
  `;

  const context = { user: owner };

  expect.assertions(6);
  const result = await request(query, { context, variables });
  expect(result).toMatchSnapshot();
  expect(result.data.unblockCommunityMember).toEqual(true);

  // ensure that only one record exists for the moderator
  const getUsersCommunities = () =>
    db
      .table('usersCommunities')
      .getAll([input.userId, input.communityId], {
        index: 'userIdAndCommunityId',
      })
      .run();

  const communityConnections = await getUsersCommunities();
  expect(communityConnections).toHaveLength(1);
  expect(communityConnections[0].isBlocked).toEqual(false);
  expect(communityConnections[0].isMember).toEqual(true);
  expect(communityConnections[0].receiveNotifications).toEqual(true);
});
