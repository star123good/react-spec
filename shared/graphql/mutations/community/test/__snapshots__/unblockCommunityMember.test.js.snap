// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should unblock a member in the community 1`] = `
Object {
  "data": Object {
    "unblockCommunityMember": true,
  },
}
`;
